General fixes for ShardedTensor op framework.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77209
* __->__ #77191

1) Add more basic validation to all ops.
2) Ensure `register_on_local_shards` uses appropriate sharding_spec.

Differential Revision: [D36292159](https://our.internmc.facebook.com/intern/diff/D36292159/)

pritamdamania87 committed:
General fixes for ShardedTensor op framework.

1) Add more basic validation to all ops.
2) Ensure `register_on_local_shards` uses appropriate sharding_spec.

Differential Revision: [D36292159](https://our.internmc.facebook.com/intern/diff/D36292159/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/_common.py:
```diff
@@ -4,6 +4,56 @@
+    if not has_distributed_tensor:
+        raise TypeError(
+            f"torch function '{op.__name__}', with args: {args} and "
+            f"kwargs: {kwargs} are called without any DistributedTensor!"
```
do we want to convey this `DistributedTensor` term? or we just use `distributed tensor`?

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/_common.py:
```diff
@@ -4,6 +4,56 @@
+                )
+            cur_pg = e.process_group
+        elif isinstance(e, ShardedTensor):
+            if cur_pg is not None and e._process_group is not cur_pg:
```
these two are essentially the same logic except that `ShardedTensor` uses `_process_group` but the other two uses `process_group`? Maybe we can consolidate this later in a separate PR to all use `_process_group`?

wanchaol reviewed (approved):
looks good!

pritamdamania87 committed:
Update on "General fixes for ShardedTensor op framework."


1) Add more basic validation to all ops.
2) Ensure `register_on_local_shards` uses appropriate sharding_spec.

Differential Revision: [D36292159](https://our.internmc.facebook.com/intern/diff/D36292159/)

[ghstack-poisoned]

fduwjj reviewed (approved):
Nice!

pritamdamania87 commented:
@pytorchbot merge this please
